### ISSUE TITLE
Sound not working on Debian Jessie

### ISSUE THREAD

**Issue by christianbundy** [raphael/linux-samus#50]
After installing the Debian packages, I restarted to find that both my touchpad and wireless card were working, which was fantastic. Unfortunately, my sound isn't working.

``` console
$ ./sound.sh
Setting sound card UCM verb
ALSA lib main.c:273:(execute_sequence) unable to open ctl device 'hw:bdwrt5677'
ALSA lib main.c:1475:(set_verb_user) error: failed to initialize new use case: HiFi
alsaucm: error failed to set _verb=HiFi: No such device
!!Failed to set UCM verb, make sure 'alsaucm' is installed
```

I thought that I might be having the same problem as #42, but was able to confirm that I'm running the correct kernel (I think).

``` console
$ uname -a
Linux cinco 4.2.4ph+ #1 SMP Fri Oct 23 16:00:38 PDT 2015 x86_64 GNU/Linux
```

It looks like the card isn't listed by ALSA either.

``` console
$ alsaucm listcards
  0: DAISY-I2S
    Daisy internal card
  1: Manta-I2S
  2: Manta-SPDIF
  3: PandaBoard
  4: PandaBoardES
  5: SDP4430
  6: Tuna
  7: apq8064-tabla-snd-card
  8: tegra-rt5640
  9: tegraalc5632
```

Checking dmesg, I found an error that I'm unsure what to do with.

``` console
$ dmesg | grep bdw
[   18.414457] bdw-rt5677 bdw-rt5677: ASoC: CPU DAI System Pin not registered
[   19.601326] bdw-rt5677 bdw-rt5677: ASoC: CPU DAI System Pin not registered
[   19.714348] bdw-rt5677 bdw-rt5677: ASoC: CPU DAI System Pin not registered
```

Covering all my bases, I checked in `/dev/snd` too.

``` console
$ ls -l /dev/snd
total 0
drwxr-xr-x  2 root root       60 Oct 24 11:01 by-path
crw-rw----+ 1 root audio 116,  2 Oct 24 11:01 controlC1
crw-rw----+ 1 root audio 116,  6 Oct 24 11:01 hwC1D0
crw-rw----+ 1 root audio 116,  3 Oct 24 11:19 pcmC1D3p
crw-rw----+ 1 root audio 116,  4 Oct 24 11:02 pcmC1D7p
crw-rw----+ 1 root audio 116,  5 Oct 24 11:02 pcmC1D8p
crw-rw----+ 1 root audio 116,  1 Oct 24 11:01 seq
crw-rw----+ 1 root audio 116, 33 Oct 24 11:01 timer
```

Thinking that this was a dupe of #13, I tried `echo "load-module module-alsa-sink device=hw:0,0" >> /etc/pulse/default.pa` but after rebooting I wasn't able to use my audio. Strangely (and potentially coincidentally), having that line added made my cursor move around erratically, even when my hand was off the touchpad. Weird, right?

Again, thanks a ton for your work on this project, and let me know if there's anything I can do to make this easier to debug. I'll submit a PR after I get this working. Cheers!

**Comment by raphael** [comment]
Can you check whether the sound drivers are loaded?
```
 ~  lsmod | grep 5677 | sort
snd_pcm                72432  8 snd_soc_rt5677,snd_soc_rt5677_spi,snd_soc_core,snd_hda_codec_hdmi,snd_soc_sst_haswell_pcm,snd_hda_codec,snd_hda_intel,snd_hda_core
snd_soc_core          124122  6 snd_soc_rt5677,snd_soc_rt5677_spi,snd_soc_sst_haswell_pcm,snd_soc_sst_bdw_rt5677_mach
snd_soc_rl6231          1882  1 snd_soc_rt5677
snd_soc_rt5677        163117  5 snd_soc_sst_bdw_rt5677_mach
snd_soc_rt5677_spi      7039  3 snd_soc_rt5677
snd_soc_sst_bdw_rt5677_mach     7785  1
snd_soc_sst_haswell_pcm    43345  4 snd_soc_sst_bdw_rt5677_mach
```

**Comment by christianbundy** [comment]
Here's what I've got.

``` console
$ lsmod | grep 5677 | sort
snd_pcm                72432  8 snd_soc_rt5677,snd_soc_rt5677_spi,snd_soc_core,snd_hda_codec_hdmi,snd_soc_sst_haswell_pcm,snd_hda_codec,snd_hda_intel,snd_hda_core
snd_soc_core          124122  6 snd_soc_rt5677,snd_soc_rt5677_spi,snd_soc_sst_haswell_pcm,snd_soc_sst_bdw_rt5677_mach
snd_soc_rl6231          1882  1 snd_soc_rt5677
snd_soc_rt5677        163117  5 snd_soc_sst_bdw_rt5677_mach
snd_soc_rt5677_spi      7039  3 snd_soc_rt5677
snd_soc_sst_bdw_rt5677_mach     7785  1 
snd_soc_sst_haswell_pcm    43345  4 snd_soc_sst_bdw_rt5677_mach
```

**Comment by raphael** [comment]
Yes that looks good which means the right kernel is running. Another thing to double check is that the firmware has been copied to the right place. Note that `alsaucm listcards` won't list the card but that's not an issue. `aplay -l` should list it though - something else to check.

**Comment by harrybraviner** [comment]
I am having the same issue as christianbundy. I get the same output as him, and doing `aplay -l` produces the output
```
**** List of PLAYBACK Hardware Devices ****
card 0: HDMI [HDA Intel HDMI], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: HDMI [HDA Intel HDMI], device 7: HDMI 1 [HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 0: HDMI [HDA Intel HDMI], device 8: HDMI 2 [HDMI 2]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```
i.e. the bdw-rt5677 is not listed.

**Comment by raphael** [comment]
Can you confirm that you copied the firmware to the right place? Maybe the issue is the location of the firmware files on Jessie is not where the script `sound.sh` copies them?

**Comment by harrybraviner** [comment]
The script copied the files to /lib/firmware/, which I'm pretty sure is the right place. All my other firmware files are there, and this debian page seem to state that it's the correct place: https://www.debian.org/releases/stable/i386/ch06s04.html.en

(Thanks for all your hard work on this, by the way - I'm very grateful for everything that does work!)

**Comment by ehegnes** [comment]
I am having the same issues as @christianbundy, however I am experiencing them on Gentoo with the 4.3 kernel. The only difference between my issues and this is that my kernel also doesn't recognize HDMI audio capabilities.

Glad to dump any output necessary to help fix this.

**Comment by ehegnes** [comment]
I was experiencing the same issue on Gentoo. After a look at #13, I realized that the firmware file `IntcSST2.bin` should be placed under an `intel` folder in your firmware directory (`/lib/firmware/intel` for me).

Frankly I'm surprised more people haven't encountered this issue. It may be worth adding an `intel` directory to `${REPO}/firmware` if this is, indeed, where the file should reside on most systems.

**Comment by raphael** [comment]
Good catch! I fixed it in master. @harrybraviner @christianbundy does moving the file `/lib/firmware/IntcSST2.bin` to `/lib/firmware/intel/IntcSST2.bin` help? (need to reboot after moving the file).

**Comment by harrybraviner** [comment]
The sound still isn't working, but I think we have some progress. `aplay -l` now shows
```
**** List of PLAYBACK Hardware Devices ****
card 0: bdwrt5677 [bdw-rt5677], device 0: System Playback (*) []
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: HDMI [HDA Intel HDMI], device 3: HDMI 0 [HDMI 0]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: HDMI [HDA Intel HDMI], device 7: HDMI 1 [HDMI 1]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
card 1: HDMI [HDA Intel HDMI], device 8: HDMI 2 [HDMI 2]
  Subdevices: 1/1
  Subdevice #0: subdevice #0
```
But I don't have any audio output. Re-running your `sound.sh` scripts produces
```
Setting sound card UCM verb
Restoring asla config
/etc/acpi/handler.sh not found. Not setting up ACPI hooks. Headphones may not work correctly!
Install the acpid daemon on re-run this script.
Sound setup completed successfully.
```
I do have the acpid package installed, but that script does not exist anywhere. If I open `alsamixer`, I can select the `bdw-rt5677` card, but the master volume slider jumps back to zero whenever I exit. I've confirmed that no audio plays even if I have `alsamixer` open and move that slider up.
Finally, I have confirmed that the drivers are loaded with `lsmod | grep 5677`:
```
snd_soc_sst_bdw_rt5677_mach     7785  2
snd_soc_sst_haswell_pcm    43345  4 snd_soc_sst_bdw_rt5677_mach
snd_soc_rt5677        163117  5 snd_soc_sst_bdw_rt5677_mach
snd_soc_rl6231          1882  1 snd_soc_rt5677
snd_soc_rt5677_spi      7039  3 snd_soc_rt5677
snd_soc_core          124122  6 snd_soc_rt5677,snd_soc_rt5677_spi,snd_soc_sst_haswell_pcm,snd_soc_sst_bdw_rt5677_mach
snd_pcm                72432  8 snd_soc_rt5677,snd_soc_rt5677_spi,snd_soc_core,snd_hda_codec_hdmi,snd_soc_sst_haswell_pcm,snd_hda_codec,snd_hda_intel,snd_hda_core
```

**Comment by ehegnes** [comment]
@harrybraviner I also did not have `/etc/acpi/handler.sh`, so I copied `/etc/acpi/default.sh` over to compensate. Probably a distro (Gentoo) difference, possibly a package version difference, definitely a dirty fix, but it worked for me.

If I cared enough, I would do some research on why I didn't have `handler.sh` to start...

**Comment by harrybraviner** [comment]
I don't have `/etc/acpid/default.sh` either! I do have `/usr/share/doc/acpid/examples/default.sh`, which I copied over. Unfortunately, this hasn't changed the result.

Do you get this behavior with `alsamixer` where it doesn't seem to pick up bdw-rt5677 as the default device?

**Comment by ehegnes** [comment]
I don't use `alsamixer`, sorry. I use `pavucontrol` with Pulseaudio (minus all that `systemd` garbage like `consolekit`).

As far the default device is concerned, I did have an issue. Initially the HDMI sink was set as default (probably because it is sink 0 on my box). I ran
```shell
pacmd list-sinks | grep -e 'name:' -e 'index:'
```
to view my sinks (the index with the star beside it is the current default sink).
And I used
```shell
pacmd set-default-sink <index>
```
to change the default to `bdw-rt5667 Analog Stereo`.
